### PR TITLE
Introduce an interface for external GC backends

### PR DESCRIPTION
**Pull request by d-netto**
Still an early draft and heavily inspired by some of the work done in the [MMTk fork of Julia.
](https://github.com/mmtk/julia)

The `gc-interface-*` files should provide a guide for folks who wish to implement an external GC backend (e.g. `MMTk`).

Ideally, we should reach a point where external GC backends can just be compiled into a separate library which simply implements the interface, and later be linked against Julia.

This could prevent inlining of functions such as write barriers used by the runtime though. A middle ground implemented for now is to define the runtime write-barriers as `STATIC_INLINE` in header files, whose implementations should be overwritten in the headers themselves when implementing an external GC backend.

LTO should solve some of these problems, but enabling it would require an additional performance assessment.

### CODE REVIEW

File: src/gc-interface-allocation.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+#define JL_SMALL_BYTE_ALIGNMENT 16
+#define JL_CACHE_BYTE_ALIGNMENT 64
+// JL_HEAP_ALIGNMENT is the maximum alignment that the GC can provide
+#define JL_HEAP_ALIGNMENT JL_SMALL_BYTE_ALIGNMENT
+#define GC_MAX_SZCLASS (2032-sizeof(void*))
+
+struct _jl_value_t *jl_gc_pool_alloc_noinline(jl_ptls_t ptls, int pool_offset, int osize);
+struct _jl_value_t *jl_gc_big_alloc_noinline(jl_ptls_t ptls, size_t allocsz);
+
+#ifndef THIRD_PARTY_GC
```
Comment (NHDaly) [inline src/gc-interface-allocation.h:23]: It seems like this file isn't really an _interface_ file - it's an implementation, right? It's implementing the interface `jl_gc_alloc`, right?

I would maybe reorganize this so that this file simply contains the *declarations* for the allocation interface, and then if not THIRD_PARTY_GC, you include the native GC header.
Something like this:
```c
JL_DLLEXPORT jl_value_t *jl_gc_alloc(jl_ptls_t ptls, size_t sz, void *ty);
uint8_t JL_CONST_FUNC jl_gc_szclass(unsigned sz) JL_NOTSAFEPOINT;
...
```
then, either in this header, or maybe even better in julia-internal.h, you'd have:
```c
#ifndef THIRD_PARTY_GC
#include "gc-native-impl-allocation.h"
#endif
```
or something like that?
Basically, following something like what is suggested here: https://stackoverflow.com/a/9317494/751061

Comment (d-netto) [reply src/gc-interface-allocation.h:23]: Even better than that IMO would be to make the file an actual interface (e.g. implementation free), as you are suggesting, and move the implementation into a separate `.c` file.

This would be the best choice for keeping MMTk or any external backend as a separate library that we simply link into Julia, as mentioned above.

IIRC, inning across different `.o` files is not possible unless you have LTO though...

Comment (NHDaly) [reply src/gc-interface-allocation.h:23]: Yeah. I think it's fine to have this be a build-time choice, not a link-time choice. I agree that would be the coolest, if you didn't have to rebuild julia, but the reality is that for inlining perf we probably will have to make it a build-time choice, yeah 👍 

You should check in with the ANU folks on these details; i think they have a lot of experience with the various options available here 😊

File: src/gc-interface-threads.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#define JL_GC_INTERFACE_THREADS_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// parallel gc thread function
+void jl_parallel_gc_threadfun(void *arg);
+// concurrent gc thread function
+void jl_concurrent_gc_threadfun(void *arg);
```
Comment (kpamnany) [inline src/gc-interface-threads.h:13]: Aren't these part of the stock GC? I don't see why they're part of the GC interface.

File: src/gc-object-layout.h [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    struct _jl_excstack_t *excstack;
+    // current exception handler
+    struct _jl_handler_t *eh;
+    // saved thread state
+    jl_ucontext_t ctx;
+    void *stkbuf; // malloc'd memory (either copybuf or stack)
+    size_t bufsz; // actual sizeof stkbuf
+    unsigned int copy_stack:31; // sizeof stack for copybuf
+    unsigned int started:1;
+} jl_task_t;
```
Comment (kpamnany) [inline src/gc-object-layout.h:188]: I don't think the task structure should be here. It is not part of GC or have anything to do with object layout.

File: src/gc-object-layout.h [hunk 4]
```diff
@@ -0,0 +1,199 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#ifndef JL_GC_OBJECT_LAYOUT_H
+#define JL_GC_OBJECT_LAYOUT_H
```
Comment (kpamnany) [inline src/gc-object-layout.h:4]: I don't see the object layout types as part of the GC subsystem and I don't think this is the right name for this file.

File: src/gc-page-profiler.h [hunk 5]
```diff
@@ -3,7 +3,7 @@
 #ifndef GC_PAGE_PROFILER_H
 #define GC_PAGE_PROFILER_H
 
-#include "gc.h"
+#include "gc-stock.h"
```
Comment (kpamnany) [inline src/gc-page-profiler.h:6]: As with the previous comment -- isn't this stock GC specific? Does it need some `#ifndef`?

File: src/gc-stock.h [hunk 6]
```diff
@@ -41,6 +42,52 @@ extern "C" {
 #define GC_PAGE_SZ (1 << GC_PAGE_LG2)
 #define GC_PAGE_OFFSET (JL_HEAP_ALIGNMENT - (sizeof(jl_taggedvalue_t) % JL_HEAP_ALIGNMENT))
 
+// =======
+// finalization
+// =======
+
+void schedule_finalization(void *o, void *f) JL_NOTSAFEPOINT;
+JL_DLLEXPORT void jl_gc_add_finalizer_th(jl_ptls_t ptls, jl_value_t *v, jl_function_t *f) JL_NOTSAFEPOINT;
+JL_DLLEXPORT void jl_finalize_th(jl_task_t *ct, jl_value_t *o);
```
Comment (kpamnany) [inline src/gc-stock.h:51]: Aren't these part of the GC interface as they're exported interfaces? Shouldn't they go to a GC interface file?

File: src/julia_gcext.h [hunk 7]
```diff
@@ -25,6 +25,9 @@ typedef void (*jl_gc_cb_post_gc_t)(int full) JL_NOTSAFEPOINT;
 typedef void (*jl_gc_cb_notify_external_alloc_t)(void *addr, size_t size) JL_NOTSAFEPOINT;
 typedef void (*jl_gc_cb_notify_external_free_t)(void *addr) JL_NOTSAFEPOINT;
 
+// Notify if system is under pressure
+typedef void (*jl_gc_cb_notify_gc_pressure_t)(void) JL_NOTSAFEPOINT;
```
Comment (kpamnany) [inline src/julia_gcext.h:29]: I think this should go in a separate PR if it is being added and not just moved from somewhere else.

File: src/llvm-final-gc-lowering.cpp [hunk 8]
```diff
@@ -225,6 +207,23 @@ void FinalLowerGC::lowerGCAllocBytes(CallInst *target, Function &F)
...
+void FinalLowerGC::lowerSafepoint(CallInst *target, Function &F)
+{
+    ++SafepointCount;
+    assert(target->arg_size() == 1);
+    IRBuilder<> builder(target);
+    Value* signal_page = target->getOperand(0);
+    builder.CreateLoad(T_size, signal_page, true);
+    target->eraseFromParent();
+}
+
```
Comment (kpamnany) [inline src/llvm-final-gc-lowering.cpp:226]: Seems like just a cosmetic change? I suggest dropping this change to simplify an already large PR.

**Review by kpamnany — CHANGES_REQUESTED** [review]
Thanks for doing this! I've requested a number of changes to make the GC code easier to find and understand. Some of the code movements aren't right (esp. moving `jl_task_t`).

File: src/llvm-final-gc-lowering.cpp (hunk 8 above)
Comment (d-netto) [reply src/llvm-final-gc-lowering.cpp:226]: It's probably a vestigial change from when I was considering changing a few things in codegen for the interface.

Will remove it.

File: src/julia_gcext.h (hunk 7 above)
Comment (d-netto) [reply src/julia_gcext.h:29]: Done. https://github.com/JuliaLang/julia/pull/52962.

File: src/Makefile [hunk 9]
```diff
@@ -44,16 +44,17 @@ SRCS := \
...
 	crc32c APInt-C processor ircode opaque_closure codegen-stubs coverage runtime_ccall
 
 RT_LLVMLINK :=
 CG_LLVMLINK :=
 
 ifeq ($(JULIACODEGEN),LLVM)
-CODEGEN_SRCS := codegen jitlayers aotcompile debuginfo disasm llvm-simdloop llvm-muladd \
-	llvm-final-gc-lowering llvm-pass-helpers llvm-late-gc-lowering llvm-ptls \
+CODEGEN_SRCS := codegen jitlayers aotcompile debuginfo \
+	disasm llvm-simdloop llvm-muladd llvm-final-gc-lowering llvm-pass-helpers llvm-late-gc-lowering llvm-ptls \
```
[the remark this answers is not in the document]
Comment (d-netto) [reply src/Makefile]: Again, probably vestigial from when I was doing codegen work. Will remove.

File: src/gc-debug.c [hunk 10]
```diff
@@ -17,6 +17,8 @@
 extern "C" {
 #endif
 
+#ifndef THIRD_PARTY_GC // third party GC backend must provide its own debugging infrastructure
```
[the remark this answers is not in the document]
Comment (d-netto) [reply src/gc-debug.c]: Resolved.

File: src/gc-page-profiler.h (hunk 5 above)
Comment (d-netto) [reply src/gc-page-profiler.h:8]: Yes. It's stock GC specific. Resolved.

**Comment by inkydragon** [comment]
Build ac22293 with debug+assert version:
```
// Make.user
FORCE_ASSERTIONS=1
WITH_GC_VERIFY:=1

// options.h
MEMDEBUG
GC_ASSERT_PARENT_VALIDITY
GC_FINAL_STATS
MEMPROFILE
GC_TIME
```

In build process, I got an error:
```
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ff996391568 -- jl_genericmemory_nbytes at D:/jl/julia/src\gc-stock.c:560
in expression starting at range.jl:163
jl_genericmemory_nbytes at D:/jl/julia/src\gc-stock.c:560
gc_stats_big_obj at D:/jl/julia/src\gc-debug.c:1138
```

<details>
<summary>full log</summary>

```
$ make
    JULIA julia-w64/usr/lib/julia/corecompiler.ji
GC mark pause 4.48 ms | scanned 0 kB = 0 + 0 | remset 0 0
0 objects (0% old), 0 kB (0% old) total allocated, 0 total fragments (0% overhead), in 0 pages

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ff996391568 -- jl_genericmemory_nbytes at D:/jl/julia/src\gc-stock.c:560
in expression starting at range.jl:163
jl_genericmemory_nbytes at D:/jl/julia/src\gc-stock.c:560
gc_stats_big_obj at D:/jl/julia/src\gc-debug.c:1138
_jl_gc_collect at D:/jl/julia/src\gc-stock.c:2655
ijl_gc_collect at D:/jl/julia/src\gc-impl-stock-collector.c:408
maybe_collect at D:/jl/julia/src\gc-impl-stock-collector.c:336
jl_gc_big_alloc_inner at D:/jl/julia/src\gc-impl-stock-collector.c:183
ijl_gc_big_alloc at D:/jl/julia/src\gc-stock.c:445
jl_gc_pool_alloc_inner at D:/jl/julia/src\gc-impl-stock-collector.c:219
jl_gc_pool_alloc_noinline at D:/jl/julia/src\gc-stock.c:681
jl_gc_alloc_ at D:/jl/julia/src\gc-impl-stock-allocation.h:132
jl_gc_alloc at D:/jl/julia/src\gc-common.c:555
_new_genericmemory_ at D:/jl/julia/src\genericmemory.c:80
jl_alloc_genericmemory at D:/jl/julia/src\genericmemory.c:123
new_array at D:/jl/julia/src\array.c:89
ijl_alloc_array_1d at D:/jl/julia/src\array.c:167
ijl_alloc_vec_any at D:/jl/julia/src\array.c:197
jl_exprn at D:/jl/julia/src\builtins.c:1479
ijl_copy_ast at D:/jl/julia/src\ast.c:963
ijl_copy_ast at D:/jl/julia/src\ast.c:967
ijl_copy_ast at D:/jl/julia/src\ast.c:967
ijl_expand_with_loc_warn at D:/jl/julia/src\ast.c:1314
jl_parse_eval_all at D:/jl/julia/src\toplevel.c:1062
ijl_load_ at D:/jl/julia/src\toplevel.c:1112
unknown function (ip: 0000023d98a91340)
jl_fptr_args at D:/jl/julia/src\gf.c:2547
_jl_invoke at D:/jl/julia/src\gf.c:2905
ijl_apply_generic at D:/jl/julia/src\gf.c:3082
unknown function (ip: 0000023d98a9219c)
jl_fptr_args at D:/jl/julia/src\gf.c:2547
_jl_invoke at D:/jl/julia/src\gf.c:2905
ijl_apply_generic at D:/jl/julia/src\gf.c:3082
jl_apply at D:/jl/julia/src\julia.h:2021
do_call at D:/jl/julia/src\interpreter.c:126
eval_value at D:/jl/julia/src\interpreter.c:223
eval_stmt_value at D:/jl/julia/src\interpreter.c:174
eval_body at D:/jl/julia/src\interpreter.c:657
jl_interpret_toplevel_thunk at D:/jl/julia/src\interpreter.c:815
jl_toplevel_eval_flex at D:/jl/julia/src\toplevel.c:941
jl_eval_module_expr at D:/jl/julia/src\toplevel.c:215
jl_toplevel_eval_flex at D:/jl/julia/src\toplevel.c:741
ijl_toplevel_eval at D:/jl/julia/src\toplevel.c:950
ijl_toplevel_eval_in at D:/jl/julia/src\toplevel.c:992
unknown function (ip: 0000023d98a90920)
jl_fptr_args at D:/jl/julia/src\gf.c:2547
_jl_invoke at D:/jl/julia/src\gf.c:2905
ijl_apply_generic at D:/jl/julia/src\gf.c:3082
jl_apply at D:/jl/julia/src\julia.h:2021
do_call at D:/jl/julia/src\interpreter.c:126
eval_value at D:/jl/julia/src\interpreter.c:223
eval_stmt_value at D:/jl/julia/src\interpreter.c:174
eval_body at D:/jl/julia/src\interpreter.c:657
jl_interpret_toplevel_thunk at D:/jl/julia/src\interpreter.c:815
jl_toplevel_eval_flex at D:/jl/julia/src\toplevel.c:941
jl_parse_eval_all at D:/jl/julia/src\toplevel.c:1065
ijl_load_ at D:/jl/julia/src\toplevel.c:1112
ijl_load at D:/jl/julia/src\toplevel.c:1125
exec_program at D:/jl/julia/src\jlapi.c:905
true_main at D:/jl/julia/src\jlapi.c:962
jl_repl_entrypoint at D:/jl/julia/src\jlapi.c:1106
jl_load_repl at D:/jl/julia/cli\loader_lib.c:569
mainCRTStartup at D:/jl/julia/cli\loader_exe.c:58
BaseThreadInitThunk at C:\WINDOWS\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
Allocations: 334642 (Pool: 0; Big: 334642); GC: 0
```
</details>

File: src/gc-object-layout.h (hunk 3 above)
Comment (d-netto) [reply src/gc-object-layout.h]: Agreed.

**Comment by d-netto** [comment]
@inkydragon: Is the failure on Windows reported above still reproducible in the latest commit?

**Comment by inkydragon** [comment]
@d-netto 

8d0f142 still gives the same error:
```
    JULIA julia-w64/usr/lib/julia/corecompiler.ji
GC mark pause 3.10 ms | scanned 0 kB = 0 + 0 | remset 0 0
0 objects (0% old), 0 kB (0% old) total allocated, 0 total fragments (0% overhead), in 0 pages

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ffc02e8155f -- jl_genericmemory_nbytes at D:/jl/julia/src\gc-stock.c:564
in expression starting at range.jl:155
jl_genericmemory_nbytes at D:/jl/julia/src\gc-stock.c:564
gc_stats_big_obj at D:/jl/julia/src\gc-debug.c:1111
```

File: src/Makefile [hunk 11]
```diff
@@ -44,16 +44,17 @@ SRCS := \
 	jltypes gf typemap smallintset ast builtins module interpreter symbol \
 	dlload sys init task array genericmemory staticdata toplevel jl_uv datatype \
 	simplevector runtime_intrinsics precompile jloptions mtarraylist \
-	threading scheduler stackwalk gc gc-debug gc-pages gc-stacks gc-alloc-profiler gc-page-profiler method \
+	threading scheduler stackwalk gc-common gc-impl-stock-collector gc-impl-stock-threads gc-stock \
+	gc-debug gc-pages gc-stacks gc-alloc-profiler gc-page-profiler method \
```
[the remark this answers is not in the document]
Comment (d-netto) [reply src/Makefile:48]: Discussed internally that we won't be doing it in this PR.

**Comment by inkydragon** [comment]
594e6ef seems to work fine now on Windows.
It can be built successfully and passed the gc test.

Build debug version with :
```
// Make.user
FORCE_ASSERTIONS=1
WITH_GC_VERIFY:=1

// options.h
MEMDEBUG
MEMFENCE
GC_ASSERT_PARENT_VALIDITY
GC_FINAL_STATS
```

----

I have a question: how to force the GC to use a single thread?
When I turn on `WITH_GC_VERIFY` I get the following warning:
```
Warn. GC verify disabled in multi-threaded GC
```

`--gcthreads=1,0` seems to not work for me.
